File: ios/EmergencyApp/Sources/RealScheduler.swift
import Foundation
import PersonStateMachine

/// `Scheduler` de producción, respaldado por `DispatchQueue.main`. El
/// `PersonStateMachine` en sí no depende de `Dispatch` — solo esta capa de
/// app, para poder mantener el módulo testeado libre de esa dependencia.
final class RealScheduler: Scheduler {
    private final class Entry: SchedulerToken {
        let workItem: DispatchWorkItem
        init(workItem: DispatchWorkItem) { self.workItem = workItem }
    }

    func schedule(after seconds: TimeInterval, _ action: @escaping () -> Void) -> SchedulerToken {
        let workItem = DispatchWorkItem(block: action)
        DispatchQueue.main.asyncAfter(deadline: .now() + seconds, execute: workItem)
        return Entry(workItem: workItem)
    }

    func cancel(_ token: SchedulerToken) {
        (token as? Entry)?.workItem.cancel()
    }
}

File: android/app/src/main/kotlin/com/farosos/app/RealScheduler.kt
package com.farosos.app

import android.os.Handler
import android.os.Looper
import com.farosos.personstate.Scheduler
import com.farosos.personstate.SchedulerToken

/**
 * `Scheduler` de producción, respaldado por el `Handler` del main looper. El
 * `PersonStateMachine` en sí no depende de Android — solo esta capa de app,
 * para poder mantener el módulo `:personstate` testeado libre de esa
 * dependencia.
 */
class RealScheduler : Scheduler {
    private class Entry(val runnable: Runnable) : SchedulerToken

    private val handler = Handler(Looper.getMainLooper())

    override fun schedule(afterSeconds: Double, action: () -> Unit): SchedulerToken {
        val runnable = Runnable { action() }
        handler.postDelayed(runnable, (afterSeconds * 1000).toLong())
        return Entry(runnable)
    }

    override fun cancel(token: SchedulerToken) {
        (token as? Entry)?.let { handler.removeCallbacks(it.runnable) }
    }
}
